### PR TITLE
Let Expression

### PR DESCRIPTION
**Pull request by HoshinoTented**

### CODE REVIEW

**Comment by HoshinoTented** [comment]
The new syntax makes Grammar-Kit unhappy, how to improve it?

**Comment by HoshinoTented** [comment]
Thanks @imkiva !!

**Comment by HoshinoTented** [comment]
The let-expression isn't a syntax sugar now, should it?

**Comment by HoshinoTented** [comment]
FINAL QUESTION:
`=>` or `:=` or others

@imkiva @ice1000 @lunalunaa

**Comment by HoshinoTented** [comment]
One problem is that we can use `let` in type (I think so does `match`). Is it allowed?

**Comment by HoshinoTented** [comment]
> One problem is that we can use `let` in type (I think so does `match`). Is it allowed?

Also, do we need to allow declaring a function by `let`, like `let f a := g a a in f x`?
For the non-recursion function, it is trivial: `let f := \ a => g a a in f x`, but the problem is the recursion function. Do we have something like `fix` in Coq?

**Review by ice1000 — COMMENTED** [review]
Can I propose another syntax? What do you think about let | a b := c d | e f := g h which is the same as partial

**Comment by HoshinoTented** [comment]
> Can I propose another syntax? What do you think about let | a b := c d | e f := g h which is the same as partial

So we should able to define functions by the `let` expression?

**Comment by ice1000** [comment]
> > Can I propose another syntax? What do you think about let | a b := c d | e f := g h which is the same as partial
> 
> So we should able to define functions by the `let` expression?

I expected function definitions, otherwise `let` is too limited. You can always use lambdas though.

**Comment by HoshinoTented** [comment]
> > > Can I propose another syntax? What do you think about let | a b := c d | e f := g h which is the same as partial
> > 
> > 
> > So we should able to define functions by the `let` expression?
> 
> I expected function definitions, otherwise `let` is too limited. You can always use lambdas though.

Then the syntax will look like:

```aya
let
  | f (x : X) : F := g x
in f y
```

which is equivalent to:

```aya
(\ (f : X -> F) => f y) (\ (x : X) => g x)
```

**Comment by ice1000** [comment]
Where did the `: F` go?

**Comment by HoshinoTented** [comment]
> Where did the `: F` go?

`f : X -> F`

**Comment by ice1000** [comment]
> > One problem is that we can use `let` in type (I think so does `match`). Is it allowed?
> 
> Also, do we need to allow declaring a function by `let`, like `let f a := g a a in f x`? For the non-recursion function, it is trivial: `let f := \ a => g a a in f x`, but the problem is the recursion function. Do we have something like `fix` in Coq?

No. I expect no recursive functions over there.

Note that we may in the future extend this functionality a little bit. Say, we can define a `let x := y` where `x` will _not_ reduce in the function body.

**Comment by ice1000** [comment]
Fanzheng Arend has implemented that. I conjecture that they did it for reasons.

**Comment by HoshinoTented** [comment]
TODO: update distiller

**Comment by ice1000** [comment]
Merge main?

File: base/src/main/java/org/aya/tyck/ExprTycker.java [hunk 1]
```diff
@@ -848,17 +849,43 @@ private Result unifyTyMaybeInsert(@NotNull Term upper, @NotNull Result result, E
     return checker.apply(term);
   }
 
-  /// region Helper
+  /**
+   * @param letBinds not empty
```
Comment (imkiva) [inline base/src/main/java/org/aya/tyck/ExprTycker.java:853]: But you actually handled the empty case, right?
I suggest remove this comment.

File: base/src/main/java/org/aya/tyck/ExprTycker.java [hunk 2]
```diff
@@ -848,17 +849,43 @@ private Result unifyTyMaybeInsert(@NotNull Term upper, @NotNull Result result, E
...
 
-  public <R> R subscoped(@NotNull Supplier<R> action) {
     var parent = this.localCtx;
     this.localCtx = parent.deriveMap();
-    var result = action.get();
+    this.localCtx.put(param);
+
+    var result = desugarLet(remain, leaf);
+
     this.localCtx = parent;
```
Comment (imkiva) [inline base/src/main/java/org/aya/tyck/ExprTycker.java:879]: `.subscoped` ?

File: base/src/main/java/org/aya/core/term/Term.java [hunk 3]
```diff
@@ -40,7 +40,7 @@
  * @author ice1000
  */
 public sealed interface Term extends AyaDocile, Restr.TermLike<Term>
-  permits Callable, CoeTerm, Elimination, ErasedTerm, FormulaTerm, HCompTerm, IntervalTerm, MatchTerm, MetaPatTerm, MetaLitTerm, PartialTerm, PiTerm, RefTerm, RefTerm.Field, SigmaTerm, StableWHNF {
+        permits Callable, CoeTerm, Elimination, ErasedTerm, FormulaTerm, HCompTerm, IntervalTerm, MatchTerm, MetaLitTerm, MetaPatTerm, PartialTerm, PiTerm, RefTerm, RefTerm.Field, SigmaTerm, StableWHNF {
```
Comment (imkiva) [inline base/src/main/java/org/aya/core/term/Term.java:43]: formats

File: base/src/test/resources/success/src/LetExpr.aya [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+  let
+  | n1 := suc n
+  | n2 := suc n1
+  in n2
+
+example def normalize : slow_plus2 7 = 9 => idp
+
+// TODO: should we allow this?
+
+def letInType : (let a := 1 in a = a) => idp
```
Comment (imkiva) [inline base/src/test/resources/success/src/LetExpr.aya:14]: Yes. They seems quite innocent.

File: cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java [hunk 5]
```diff
@@ -792,6 +800,17 @@ private Arg<Pattern> unitPattern(@NotNull GenericNode<?> node) {
     return Expr.Array.newList(entireSourcePos, exprs);
   }
 
+  public @NotNull Expr.Let.Bind letBind(@NotNull GenericNode<?> node) {
+    var pos = sourcePosOf(node);
+    var bind = weakId(node.child(WEAK_ID));
+    // make IDEA happy
+    var teles = lambdaTelescope((SeqView<GenericNode<?>>) node.childrenOfType(LAMBDA_TELE));
```
Comment (imkiva) [inline cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java:807]: `node.childrenOfType(...).map(x -> x)`

File: base/src/main/java/org/aya/tyck/ExprTycker.java (hunk 1 above)
Comment (HoshinoTented) [reply base/src/main/java/org/aya/tyck/ExprTycker.java:853]: My bad

File: cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java (hunk 5 above)
Comment (HoshinoTented) [reply cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java:807]: My stupid brain rejected this code before. 😭

File: base/src/main/java/org/aya/core/term/Term.java (hunk 3 above)
Comment (HoshinoTented) [reply base/src/main/java/org/aya/core/term/Term.java:43]: 😭

File: cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java (hunk 5 above)
Comment (HoshinoTented) [reply cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java]: My stupid brain said the `.map(x -> x)` is inefficient.

Comment (ice1000) [reply cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java]: It is but I don't think it matters that much

**Review by ice1000 — APPROVED** [review]

File: cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java (hunk 5 above)
Comment (HoshinoTented) [reply cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java]: > It is but I don't think it matters that much

Yes, my bad habit.

Comment (imkiva) [reply cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java]: I think `.map` is better as it only causes compile errors, while type cast may throw runtime exceptions.

Comment (HoshinoTented) [reply cli/src/main/java/org/aya/cli/parse/AyaGKProducer.java]: > I think `.map` is better as it only causes compile errors, while type cast may throw runtime exceptions.

YES! I got it.

**Comment by HoshinoTented** [comment]
bors d=imkiva

**Review by imkiva — DISMISSED** [review]
bors d=HoshinoTented

**Comment by imkiva** [comment]
bors merge

**Comment by ice1000** [comment]
> Merge conflict.

Why

**Comment by HoshinoTented** [comment]
> > Merge conflict.
> 
> Why

Parser

**Comment by ice1000** [comment]
bors merge

**Pull request merged**